### PR TITLE
Spark : Support parallelism in RemoveOrphanFiles

### LINKED ISSUE #3582: Delete File Performance in ExpireSnapshots and RemoveOrphans
For ExpireSnapshotAction and RemoveOrphanFile, if there are a lot of files to remove then the Spark jobs run very slow (can timeout).  

I guess current design for driver-side delete is to be prevent throttling or DDOS'ing the storage system, but I wonder if this can be made flexible.  

- Could we make a configuration to allow deletion on the Spark executors?
- Could we make a batch deleteObjects() interface in FileIO to take advantage of some faster delete methods native to different storages?  For example AWS Multi-Object Delete (1000 at a time)

### PR DESCRIPTION
**Pull request by dramaticlly**
# Notes

Allow parallelism in removing orphan files using Spark procedures and Spark Actions, like how it was implemented in expire snapshots, This change close #3582 

Since I need to change the `DeleteOrphanFiles` interface in api, ends up also updating the code for spark2.4/3.0/3.1/3.2 BaseDeleteOrphanFilesSparkAction as they are implement the same interface, also update the spark procedures to allow calling from spark SQL.

## Tests

I update the tests for `TestRemoveOrphanFilesAction` and `TestRemoveOrphanFilesProcedure` in respective spark version. I did not touch any tests in  `TestRemoveOrphanFilesAction3` as it mostly cover different catalogs

```
$ ./gradlew -DsparkVersions=3.2 \
-DhiveVersions= -DflinkVersions= \
:iceberg-spark:iceberg-spark-3.2_2.12:check \
:iceberg-spark:iceberg-spark-extensions-3.2_2.12:check \
:iceberg-spark:iceberg-spark-runtime-3.2_2.12:check \
-Pquick=true -x javadoc

BUILD SUCCESSFUL in 29s
54 actionable tasks: 41 executed, 13 up-to-date

```

### CODE REVIEW

**Comment by dramaticlly** [comment]
@szehon-ho  and @RussellSpitzer , can you help kicked of the CI workflow for me as first time contributor? Feel free to take a look and I can do more local testing in the meantime. Thank you!

**Comment by RussellSpitzer** [comment]
Approved to run

**Comment by dramaticlly** [comment]
Updated Spark 2.4/3.0/3.1 Actions due to interface change in api, somehow it was not reflected in IDE as those spark version was not added as src. 

anyway I did run the check and test for each spark version and they are all passing, and also rebased with latest upstream. Hopefully CI would be happy now

**Comment by dramaticlly** [comment]
Sorry, I think I did some rebase from upstream and fi, ends up re-require the approval in order to run the CI.  But now this PR it's ready for review and I was wondering if @RussellSpitzer can help take a look? 

Add @szehon-ho as well for issue #3582

**Comment by RussellSpitzer** [comment]
Sorry I've only been able to approve the ci, busy Monday and Tuesday!

**Comment by dramaticlly** [comment]
> Sorry I've only been able to approve the ci, busy Monday and Tuesday!

No worries, please take your time and I am happy to see all checks are passing

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java
Comment (RussellSpitzer) [inline spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java:89]: nit I would have a new line before this

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java
Comment (RussellSpitzer) [inline spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java:103]: nit I would have a new line before this

File: spark/v3.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRemoveOrphanFilesProcedure.java
Comment (RussellSpitzer) [inline spark/v3.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRemoveOrphanFilesProcedure.java:293]: We usually add these tests to `testInvalid***Cases()`

**Comment by dramaticlly** [comment]
> This looks good to me, Small request though 
> 
> 
> 
> Can we just put the changes for Spark 3.2 in this pr then the backports in another pr?

Thanks Russel, that's my original intention as well. But since I am change the interface in API, change the spark 3.2 alone but not other Spark versions will break the CI workflow. Or do you want me to separate these change out in 2 separate commits but in same PR? 

Maybe I understand you wrong, If you can elaborate a bit more in detail, I am happy to do it!

**Comment by RussellSpitzer** [comment]
Ah I totally forgot, I think it's fine then. It just makes it harder for me to check all the changes. I'll merge after we  fix up the few spacing nits.

**Comment by dramaticlly** [comment]
@RussellSpitzer, I added newline as suggested by you for baseAction for spark 2.4/3.0/3.1/3.2 and confirmed that they are all the same before and after my change, I hope that helps! 

Also looks like CI still need your approval to run for the spacing change

```
➜  iceberg git:(master) diff spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java
➜  iceberg git:(master) diff spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java
➜  iceberg git:(master) diff spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java
```

**Comment by RussellSpitzer** [comment]
Yep! Clicked it off, you will have full auth to run any CI after we merge this so you won't need me anymore :D

File: spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java
Comment (szehon-ho) [inline spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java:90]: Not sure I see a huge point to this variable (feel like a comment could have sufficed), but I guess it's there in the other version, so no matter :)

I guess the explicit assumption is that Tasks execute without a service, executes the task on a single thread.

**Review by szehon-ho — APPROVED** [review]
Looks great to me

**Comment by RussellSpitzer** [comment]
Thanks @dramaticlly for the PR and @szehon-ho for the review! Merged!

**Pull request merged**
